### PR TITLE
correction

### PR DESCRIPTION
**Pull request by tanjiarui**
the original version doesn't make sense

### CODE REVIEW

File: losses/detr_loss.py [hunk 1]
```diff
@@ -1,104 +1,85 @@
...
+
+		# cost weights
+		self.cost_class_weight = cost_class_weight
+		self.cost_box_weight = cost_box_weight
+		self.cost_giou_weight = cost_giou_weight
+
+		# define necessary losses
+		self.regression_loss = tf.keras.losses.MeanAbsoluteError()
+		self.generalized_iou_loss = tfa.losses.GIoULoss(mode='giou')
+		self.classification_loss = tf.keras.losses.CategoricalCrossentropy()
```
Comment (AlexanderBogatko) [inline losses/detr_loss.py]: Good way, but in original paper authors use weight=0.1 for `background` class to avoid imbalance, so it was the reason why i implement `Weightened` version of standard `CategoricalCrossentropy` loss.

File: losses/detr_loss.py [hunk 2]
```diff
@@ -1,104 +1,85 @@
...
+		self.num_classes = num_classes
+
+		# cost weights
+		self.cost_class_weight = cost_class_weight
+		self.cost_box_weight = cost_box_weight
+		self.cost_giou_weight = cost_giou_weight
+
+		# define necessary losses
+		self.regression_loss = tf.keras.losses.MeanAbsoluteError()
+		self.generalized_iou_loss = tfa.losses.GIoULoss(mode='giou')
```
Comment (AlexanderBogatko) [inline losses/detr_loss.py]: You can't use `GIoULoss` as is, because it expect boxes in corner form (see it in docstring here: https://github.com/tensorflow/addons/blob/v0.17.0/tensorflow_addons/losses/giou_loss.py#L75), but DeTR outputs boxes in center form (see docstring, line 40 and original paper). So you need an adapter to connect this two entities.

File: losses/detr_loss.py [hunk 3]
```diff
@@ -1,104 +1,85 @@
...
+		# perform matching algorithm for each sample in batch
+		total_loss = tf.constant(0, dtype=tf.float32)
+		for batch_id in range(tf.shape(gt_boxes)[0]):
+			# construct cost matrix
+			cost_class = -tf.gather(predicted_scores[batch_id], gt_labels[batch_id], axis=1)  # (N[predictions], N[target])
+			# gt_boxes_expanded = gt_boxes[batch_id][tf.newaxis, ...]  # (1, N, 4)
+			# predicted_boxes_expanded = gt_boxes[batch_id][:, tf.newaxis, :]  # (N, 1, 4)
+
+			# (N[predictions], N[target])
+			cost_boxes = self.regression_loss(y_true=gt_boxes[batch_id], y_pred=predicted_boxes[batch_id])
```
Comment (AlexanderBogatko) [inline losses/detr_loss.py]: This is not correct, because you don't get `(N[predictions], N[target])` shape after this operation, pls, use `_expanded` versions of ground-truth and predicted boxes.

File: losses/detr_loss.py [hunk 4]
```diff
@@ -1,104 +1,85 @@
...
+			cost_generalized_iou = self.generalized_iou_loss(y_true=gt_boxes[batch_id], y_pred=predicted_boxes[batch_id])  # (N[preds], N[gt])
```
Comment (AlexanderBogatko) [inline losses/detr_loss.py]: Breaks PEP8 standard about length of single line.

File: losses/detr_loss.py [hunk 5]
```diff
@@ -1,104 +1,85 @@
...
+			cost_matrix = self.cost_box_weight * cost_boxes + self.cost_class_weight * cost_class + self.cost_giou_weight * cost_generalized_iou
```
Comment (AlexanderBogatko) [inline losses/detr_loss.py:126]: Breaks PEP8 standard about length of single line.

File: losses/detr_loss.py [hunk 6]
```diff
@@ -1,104 +1,85 @@
...
+			positive_mask = gt_labels[batch_id] != self.background_id
+			sample_gt_boxes = tf.boolean_mask(sample_gt_boxes, positive_mask)
+			sample_pred_boxes = tf.boolean_mask(sample_pred_boxes, positive_mask)
+
+			# calculate loss values
+			class_loss = self.classification_loss(y_true=sample_gt_scores, y_pred=sample_pred_scores)
+			box_l1_loss = tf.math.reduce_mean(self.regression_loss(y_true=sample_gt_boxes, y_pred=sample_pred_boxes))
+			box_giou_loss = tf.math.reduce_mean(self.generalized_iou_loss(y_true=sample_gt_boxes, y_pred=sample_pred_boxes))
+
+			total_loss += self.cost_class_weight * class_loss + self.cost_box_weight * box_l1_loss + self.cost_giou_weight * box_giou_loss
```
Comment (AlexanderBogatko) [inline losses/detr_loss.py:149]: Breaks PEP8 standard about length of single line.

File: losses/detr_loss.py (hunk 2 above)
Comment (tanjiarui) [reply losses/detr_loss.py]: now i think it's connected

**Comment by tanjiarui** [comment]
the initial loss raised after i modified defects. does it make sense? also, i got this error when i ran in the autograph mode. looks like a data type error. any idea?
> Type inference failed. This indicates an invalid graph that escaped type checking. Error message: INVALID_ARGUMENT: expected compatible input types, but input 1:
type_id: TFT_OPTIONAL
args {
  type_id: TFT_PRODUCT
  args {
    type_id: TFT_TENSOR
    args {
      type_id: TFT_BOOL
    }
  }
}
 is neither a subtype nor a supertype of the combined inputs preceding it:
type_id: TFT_OPTIONAL
args {
  type_id: TFT_PRODUCT
  args {
    type_id: TFT_TENSOR
    args {
      type_id: TFT_LEGACY_VARIANT
    }
  }
}
while inferring type of node 'detr_loss/mean_absolute_error_5/cond/output/_37'
